tree/treeplayer/src/TDFUtils.cxx

```diff
@@ -41,60 +41,36 @@ TIgnoreErrorLevelRAII::~TIgnoreErrorLevelRAII()
 std::string
 ColumnName2ColumnTypeName(const std::string &colName, TTree *tree, TCustomColumnBase *tmpBranch, TDataSource *ds)
 {
+   // if this is a TDataSource column, we just ask the type name to the data-source
    if (ds && ds->HasColumn(colName)) {
       return ds->GetTypeName(colName);
    }
 
    TBranch *branch = nullptr;
    if (tree)
       branch = tree->GetBranch(colName.c_str());
-   if (!branch && !tmpBranch) {
-      throw std::runtime_error("Column \"" + colName + "\" is not in a file and has not been defined.");
-   }
    if (branch) {
       // this must be a real TTree branch
       static const TClassRef tbranchelRef("TBranchElement");
       if (branch->InheritsFrom(tbranchelRef)) {
          return static_cast<TBranchElement *>(branch)->GetClassName();
       } else { // Try the fundamental type
-         std::string_view title(branch->GetTitle());
-         char typeCode = title.back();
-         std::string type;
-         if (typeCode == 'B')
-            type = "char";
-         else if (typeCode == 'b')
-            type = "unsigned char";
-         else if (typeCode == 'I')
-            type = "int";
-         else if (typeCode == 'i')
-            type = "unsigned int";
-         else if (typeCode == 'S')
-            type = "short";
-         else if (typeCode == 's')
-            type = "unsigned short";
-         else if (typeCode == 'D')
-            type = "double";
-         else if (typeCode == 'F')
-            type = "float";
-         else if (typeCode == 'n')
-            type = "Float_t";
-         else if (typeCode == 'L')
-            type = "Long64_t";
-         else if (typeCode == 'l')
-            type = "ULong64_t";
-         else if (typeCode == 'O')
-            type = "bool";
-         if (type.empty())
-            throw std::runtime_error("could not deduce type of branch " + std::string(title));
-         if (title.size() > 3 && title[title.size() - 3] == ']') {
-            // title has the form "varname[size]/X", i.e. it refers to an array (doesn't matter if size is fixed or not)
-            // TDataFrame reads it as a TArrayBranch
-            return "ROOT::Experimental::TDF::TArrayBranch<" + type + ">";
+         const auto listOfLeaves = branch->GetListOfLeaves();
+         const auto nLeaves = listOfLeaves->GetEntries();
+         if (nLeaves != 1)
+            throw std::runtime_error("TTree branch " + colName + " has " + std::to_string(nLeaves) +
+                                     " leaves. Only one leaf per branch is supported.");
+         const std::string branchType = static_cast<TLeaf *>(listOfLeaves->UncheckedAt(0))->GetTypeName();
+         if (branchType.empty())
+            throw std::runtime_error("could not deduce type of branch " + std::string(colName));
+         if (static_cast<TLeaf*>(listOfLeaves->UncheckedAt(0))->GetNdata() > 1) {
```

Comment:
So here you meant.
```
TLeaf *leaf = static_cast<TLeaf*>(listOfLeaves->UncheckedAt(0));
if (leaf->GetLeafCounter() || leaf->GetStaticLen() > 1) {
```

Reply:
uh...`GetLeafCounter` was not mentioned so far.
please disregard this implementation, I will change it with the snippet discussed above